File: 2021/day1/day1.c
#include <stdio.h>
#include <stdlib.h>

#define FILE_NAME "in.txt"
#define MAX_LINE_LEN 6
#define INPUT_LEN 2000

void readInput(int* lines) {
  FILE* file = fopen(FILE_NAME, "r");
  
  for (int i = 0; i < INPUT_LEN; i++) {
    char buffer[MAX_LINE_LEN];

    if (fgets(buffer, MAX_LINE_LEN, file) == NULL) {
      fprintf(stderr, "Could not read input\n");
      exit(EXIT_FAILURE);
    }

    int depth = atoi(buffer);
    if (depth == 0) {
      fprintf(stderr, "Could not convert line to integer, at i = %d\n", i);
      exit(EXIT_FAILURE);
    }

    lines[i] = depth;
  } 

  if (fclose(file) == EOF) {
    fprintf(stderr, "could not close file\n");
    exit(EXIT_FAILURE);
  }
}

int part1(int* lines) {
  int counter = 0;

  for (int i = 1; i < INPUT_LEN; i++) {
    if (lines[i] > lines[i-1]) {
      ++counter;
    }
  }

  return counter;
}

int part2(int* lines) {
  int counter = 0;

  for (int i = 1; i < INPUT_LEN - 2; i++) {
    int prev = lines[i-1] + lines[i] + lines[i+1];
    int current = lines[i] + lines[i+1] + lines[i+2];

    if (current > prev) {
      ++counter;
    }
  }

  return counter;
}

int main() {
  int lines[INPUT_LEN];
  readInput(lines);
  
  printf("part1: %d\n", part1(lines));
  printf("part2: %d\n", part2(lines));

  return EXIT_SUCCESS;
}

File: 2020/day1/day1.py
def part1(nums: set[int]) -> int:
    for a in nums:
        b = 2020 - a
        if b in nums:
            return a * b


def part2(nums: set[int]) -> int:
    as_list = list(nums)

    for i, a in enumerate(as_list[:-1]):
        for b in as_list[i+1:]:
            c = 2020 - a - b
            if c in nums:
                return a * b * c


def main():
    filename = 'in.txt'
    with open(filename) as file:
        nums = {int(line) for line in file}
    
    print('part 1:', part1(nums))
    print('part 2:', part2(nums))


if __name__ == '__main__':
    main()
